Bluetooth: Controller: Make Extended Advertising DID unique

cvinayak opened the pull request:
Make the Extended Advertising DID value updated to be unique
for every new advertising set created.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Make Extended Advertising DID unique

Make the Extended Advertising DID value updated to be unique
for every new advertising set created.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -782,13 +782,15 @@ uint8_t ull_adv_aux_hdr_set_clear(struct ll_adv_set *adv,
 			did = sys_le16_to_cpu(pri_adi_prev->did);
+			did++;
 		} else {
+			static uint16_t did_unique;
+
 			pri_adi->sid = adv->sid;
 			sec_adi->sid = adv->sid;
+			did = did_unique++;
```
How does this ensure uniqueness? Looks to me like it is still possible that the "old" `did` value and this `did_unique` value can still end up as the same value, as they are incremented separately. 

e.g. first iteration `did_unique is 0` (statically declared) and `did` is set to 0. Next iteration if `pri_hdr_prev.adi` is true, `did` is incremented to 1. If the next iteration has `pri_hdr_prev.adi` evaluated to false, then `did` is set to 1 again. 

Furthermore, this also doesn't seem to take this requirement into account: 

> The Advertising DID for a given advertising set or periodic advertising train shall be initialized with a randomly chosen value.

cvinayak replied:
`did_unique` value is used and post incremented only when an advertising set is created, i.e. Advertising Data Information is added to the extended common header of the PDU. ADI is not removed from the advertising until the set is removed/cleared. 
Hence, for a given set (SID) the DID in the PDU is privately incremented when the PDU is updated, and for new sets unique values from `did_unique` get used.
But yes, an advertising set starting with DID as `0` from `did_unique` after an update would have `1`, and if removed, and added back would get `1` from  `did_unique`.

Maybe a simple running value (initially a random number at power up) across all sets would avoid repeated `DID` ?

Thalley replied:
> Maybe a simple running value (initially a random number at power up) across all sets would avoid repeated DID ?

That sounds like a solution that would ensure DID uniqueness as well as conforming to the other requirement (which I guess cannot be tested by any conformance test).

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -777,34 +784,42 @@ uint8_t ull_adv_aux_hdr_set_clear(struct ll_adv_set *adv,
 		}
 
-		did++;
+		/* NOTE: Using a running did_unique is a simple solution to
+		 *       prevent using same DID when PDU fields change.
+		 *       This does not prevent a case wherein an advertising
+		 *       set is removed and added again while did_unique was
+		 *       incremented by other advertising sets, and there is a
+		 *       chance the new advertising set may end up with DID
+		 *       that was used before the set was removed. This is a
+		 *       rare case, which we are not handling.
+		 */
```
I suppose the reason why we are not handling this, is because the solution to that would require a loop of all existing sets each time we increment the DID, which can be fairly expensive, and that the consequence of having two advertiser advertising with the same DID is not that bad? 

I don't know what the value is used for, so I don't know what the consequence of that is. 

In the rare case that DID is not unique, will it be non-unique for the lifetime of the advertising set, or just for the next advertising event? In order words, if it is not unique, will this be fixed over time automatically?

cvinayak replied:
> I suppose the reason why we are not handling this, is because the solution to that would require a loop of all existing sets each time we increment the DID, which can be fairly expensive, and that the consequence of having two advertiser advertising with the same DID is not that bad?

A correct solution option is, maintain an array of last used DIDs for each of maximum 16 SIDs that an application can use to create a set and use these values to set unique DID for PDU updates in a advertising set. Only expensive thing I can see is the setup time due to the requirement of 32 bytes of random value which depend on environmental conditions on how fast they are available.

I will post an update and we can review it.

Thalley replied:
> I will post an update and we can review it.

Please do :) 

> If a scanner is using duplicate filter, then if will drop PDUs with same DID from being reported to Host.

That sounds like something that can, and should be, found during conformance tests, and as such is something we probably need to ensure doesn't happen.

cvinayak replied:
> That sounds like something that can, and should be, found during conformance tests, and as such is something we probably need to ensure doesn't happen.

Haven't observed that there was any conformance test checking duplicate filters, or the tests passed as old implementation did increment DID in normal cases.

cvinayak committed:
Bluetooth: Controller: Remember last used DID for each SID

Remember last used DID for each SID, so that DID value used,
when updating the advertising PDU fields, change between
consecutive changes to PDU fields.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Explicitly check previous auxiliary PDU for ADI

To be consistent in how order header flags are handled when
updating the common header format fields, explicit check
for ADI flag in previous auxiliary PDU buffer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -285,6 +285,8 @@ enum pdu_adv_mode {
 	EXT_ADV_MODE_NON_CONN_SCAN = 0x02,
 };
 
+#define PDU_ADV_SID_COUNT 16
```
We also have the `BT_GAP_SID_MAX` define; not sure if it makes sense to use that in the controller.

cvinayak replied:
I did notice that, which was not a count but the maximum value, also it is in the GAP namespace.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -775,42 +784,29 @@ uint8_t ull_adv_aux_hdr_set_clear(struct ll_adv_set *adv,
+	sec_adi->sid = adv->sid;
 
-		did++;
+	did = did_unique[adv->sid]++;
```
So it is the DID of a SID that needs to be unique? I.e. a set with SID = 0 and a set with SID = 1 can both have DID = 2 and be compliant? 

I suggest we add a description here stating that, as well as stating that this value is in fact 12 bits and may overflow without any negative consequences, e.g. something like


```suggestion
	/* The DID for a specific SID shall be unique. 
	 * The DID is 12 bits and may overflow without any negative consequences 
	 */
	did = did_unique[adv->sid]++;
```
